sources/engine/Xenko.Voxels/Voxels/Voxelization/VoxelStorage/VoxelStorageTextureClipmap.cs

```diff
@@ -0,0 +1,192 @@
+﻿using System;
+using System.Collections.Generic;
+using Xenko.Core.Mathematics;
+using Xenko.Graphics;
+using Xenko.Rendering.Shadows;
+using Xenko.Shaders;
+
+namespace Xenko.Rendering.Voxels
+{
+    public class VoxelStorageTextureClipmap : IVoxelStorageTexture
+    {
+        public Vector3 ClipMapResolution;
+        public int ClipMapCount;
+        public int LayoutSize;
+        public float VoxelSize;
+        public Int3 VolumeTranslation;
+
+        public bool DownsampleFinerClipMaps;
+
+        public Xenko.Graphics.Texture ClipMaps = null;
+        public Xenko.Graphics.Texture MipMaps = null;
+        public Xenko.Graphics.Texture[] TempMipMaps = null;
+
+        public Vector4[] PerMapOffsetScale = new Vector4[20];
+        public Vector4[] PerMapOffsetScaleCurrent = new Vector4[20];
+        public Vector3[] MippingOffset = new Vector3[20];
+
+        ShaderClassSource sampler = new ShaderClassSource("VoxelStorageTextureClipmapShader");
+
+        public void UpdateVoxelizationLayout(string compositionName)
+        {
+
+        }
+        public void ApplyVoxelizationParameters(ObjectParameterKey<Texture> MainKey, ParameterCollection parameters)
+        {
+            parameters.Set(MainKey, ClipMaps);
+        }
+        string curMipMapShader = "";
+        Xenko.Rendering.ComputeEffect.ComputeEffectShader VoxelMipmapSimple;
+        //Memory leaks if the ThreadGroupCounts/Numbers changes (I suppose due to recompiles...?)
+        //so instead cache them as seperate shaders.
+        Xenko.Rendering.ComputeEffect.ComputeEffectShader[] VoxelMipmapSimpleGroups;
+
+        Int3 ToInt3(Vector3 v)
+        {
+            return new Int3((int)v.X, (int)v.Y, (int)v.Z);
```

Comment:
Ah to be honest I tried to avoid touching any Xenko code outside of what I had to, alright though. I'm hesitant about putting Mod in since I imagine most would expect it to behave like %, however if it's alright as long as it has a clear summary, will do.